proposals/20201026-creating-control-plane.md

```diff
@@ -0,0 +1,433 @@
+
+- Define a standard process of creating nested components for NCP.
+
+- Present two examples of creating NCP natively or using the third-party provider to create the NEtcd.
+
+### Non-Goals
+
+- Define how NCP controller works.
+
+- Discuss the details of how the third-party component provider work.
+
+## Proposal
+
+### Portability and Customizability
+
+Generally, creating the three major components requires similar high-level information, like the components' version, the number of replicas, and the amount of computing resources. Meanwhile, end-users should be able to customize NCP components, i.e., specifying the component image, version, and command-line options. Therefore, we define a new struct `NestedComponentSpec` that contains common information required by different providers as well as customized information specified by the end-users. The `NestedComponentSpec` will look like the following
+
+```go
+type NestedComponentSpec struct {
+    // ComponentSpec defines the common information for creating the component
+    // +optional
+    CommonSpec addonv1alpha1.CommonSpec `json:"commonSpec",omitempty`
+    
+    // Resources defines the amount of computing resources that will be used by this component
+    // +optional
+    Resources corev1.ResourceRequirements `json:"resources",omitempty`
+    
+    // Replicas defines the number of replicas in the component's workload 
+    // +optional
+    Replicas int32 `json:"replicas",omitempty`
+
+    // PatchSpecs includes the user specifed settings
+    // +optional
+    PatchSpec addonv1alpha1.PatchSpec `json:"patchSpec,omitempty"`
+}
+```
+
+The `CommonSpecs` and the `PatchSpec` are defined in [kubebuilder-declarative-pattern](https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/blob/1cbf859290cab81ae8e73fc5caebe792280175d1/pkg/patterns/addon/pkg/apis/v1alpha1/common_types.go):
+
+
+```go
+// CommonSpec defines the set of configuration attributes that must be exposed on all addons.
+type CommonSpec struct {
+	// Version specifies the exact addon version to be deployed, eg 1.2.3
+	// It should not be specified if Channel is specified
+	Version string `json:"version,omitempty"`
+	// Channel specifies a channel that can be used to resolve a specific addon, eg: stable
+	// It will be ignored if Version is specified
+	Channel string `json:"channel,omitempty"`
+}
+
+// +k8s:deepcopy-gen=true
+type PatchSpec struct {
+	Patches []*runtime.RawExtension `json:"patches,omitempty"`
+}
+```
+
+If the designated component's providers do not allow patching, the `PatchSpec` will be ignored. All Nested CRs need to contain a `NestedComponentSpec`. We will not discuss how NCPT and NCP controller generates the `NestedComponentSpec` in this proposal. However, a possible way is to add three fields to the NCPT CRD, and has each of them holds the `NestedComponentSpec` for one of the components. The NCPT will look like the following
```

Comment:
We can also use webhook validation to disallow certain fields to be filled, so that data doesn't get ignored silently and feedback is given to users immediately

Reply:
Yeah, this would end up being up to the controller re-implementer to be responsible for as well which makes CAPNs responsibility here significantly lessened.